### ISSUE TITLE
MUC/Sub : Error creating room using subscribe 

### ISSUE THREAD

**Issue by gsojitra** [processone/ejabberd#1404]
> What version of ejabberd are you using?

16.09

> What operating system (version) are you using?

Cent OS 7.2.1511

> How did you install ejabberd (source, package, distribution)?

source

> What did not work as expected? Are there error messages in the log? What

If subscription request is sent to non-existing room, new room will be created but subscription will fail with error "Subscriptions are not allowed".

> was the unexpected behavior? What was the expected result?

If subscription is failed with error then room also should not be created.
or
If room is created via subscription then Allow Subscription should be true for that room.

**Comment by badlop** [comment]
I can reproduce that problem in ejabberd 16.09 myself. Then in 16.12-beta1 the problem seems solved:
```
12:33:39.849 [info] Created MUC room sala1@conference.localhost by user2@localhost/tka1
12:33:39.851 [info] Destroyed MUC room sala1@conference.localhost because it's temporary and empty
12:33:39.851 [info] Stopping MUC room sala1@conference.localhost
```
Looking in the changelogs  I don't see when was it fixed... git log 16.09..16.12-beta1 --oneline src/mod_muc_room.erl

Anyway, can you test in 16.12-beta1? You can download a binary installer in another machine, just for testing this issue and confirm if it was closed, or requires more investigation.

**Comment by gsojitra** [comment]
I installed 16.12-beta1 version on Windows machine and I am able to reproduce the issue.Room is created but user is getting error "Subscriptions are not allowed".

I have attached request and response below:
Client to Server:

<iq to='t1@conference.test.com'
     type='set'
     id='E6E10350-76CF-40C6-B91B-1EA08C332FC7'>
  <subscribe xmlns='urn:xmpp:mucsub:0'
              nick='mynick'>
    <event node='urn:xmpp:mucsub:nodes:messages' />
    <event node='urn:xmpp:mucsub:nodes:affiliations' />
    <event node='urn:xmpp:mucsub:nodes:presence' />
    <event node='urn:xmpp:mucsub:nodes:subject' />
    <event node='urn:xmpp:mucsub:nodes:config' />
  </subscribe>
</iq>

Server to Client :

<iq xml:lang='en' to='admin@test.com/r1' from='t1@conference.test.com' type='error' id='E6E10350-76CF-40C6-B91B-1EA08C332FC7'>
<subscribe xmlns='urn:xmpp:mucsub:0' nick='mynick'>
    <event node='urn:xmpp:mucsub:nodes:messages'/>
    <event node='urn:xmpp:mucsub:nodes:affiliations'/>
    <event node='urn:xmpp:mucsub:nodes:presence'/>
    <event node='urn:xmpp:mucsub:nodes:subject'/>
    <event node='urn:xmpp:mucsub:nodes:config'/>
  </subscribe>
<error code='405' type='cancel'>
<not-allowed xmlns='urn:ietf:params:xml:ns:xmpp-stanzas'/>
<text xml:lang='en' xmlns='urn:ietf:params:xml:ns:xmpp-stanzas'>Subscriptions are not allowed</text>
</error>
</iq>

**Comment by zinid** [comment]
Try to set `allow_subscription: true` in the default room options, e.g.
```
mod_muc:
   ...
   default_room_options:
      ...
      allow_subscription: true
      ...
...
```

**Comment by gsojitra** [comment]
If we set `allow_subscription: true` ,it will work and i am aware about that.But the issue is if we have not set the option it will create room and then give error "not allowed".

**Comment by badlop** [comment]
With default config file, and sending your query, I see in the logs:
```
12:03:07.663 [info] Created MUC room t1@conference.localhost by user2@localhost/tka1
12:03:07.664 [info] Destroyed MUC room t1@conference.localhost because it's temporary and empty
12:03:07.664 [info] Stopping MUC room t1@conference.localhost
```
In your case, have you modified the mod_muc section in the ejabberd configuration file? Specifically, have you set persistent: true?

**Comment by gsojitra** [comment]
Correct badlop. With default config we will get above logs.
My point is that client will get error stating that "Subscriptions are not allowed" even though room is creared by him(owner).

**Comment by badlop** [comment]
Ah, you are suggesting that, when a room is created by subscribing to a new room (instead of joining the room), then the allow_subscriptions room option should become enabled, thus allowing the room to exist. Is that it?

**Comment by gsojitra** [comment]
Correct.
I can not say exactly what should be the behavior.
Enabling option on request may not be good idea since subscriber may not be allowed to change room option.
As per my understanding, we can check allow_subscriptions option before creating room.

**Comment by zuandre** [comment]
hey did you get this sorted

**Comment by badlop** [comment]
I committed two fixes to git, so it should be fixed now in the git version. It will be included in ejabberd 17.01 when released. Download and compile ejabberd git sources, or apply the fixes to your source, or wait some days for 17.01 to be released.
